### ISSUE TITLE
DOtherSide.lib

### ISSUE THREAD

**Issue by bubnenkoff** [filcuc/DOtherSide#26]
Could you provide ready to use DOtherSide.lib lib for Windows?

**Comment by filcuc** [comment]
i may, but should i? really there's many things to consider:
1) multiple compilers (mingw, visual studio, clang)
2) multiple qt version
3) multiple architecture
what i see it's that i could finish in doing
1) libDOtherSide for VS 2013 - Qt 5.5 both 64/32bit 
2) libDOtherSide for VS 2015 - Qt.5.5 both 64/32bit
3) libDOtherSide for mingw48/49 - QT.5.5 both 64/32
the result is 6 build configurations to maintain.

Furthermore this implies that i must to keep up the binaries at each compiler or qt update..dunno.
However can you shade some light what's the problem of building it from source?
This is the 2nd issue that's get opened for requesting pre-built binaries for Windows.

Given that, i made a video for showing how to build this lib for windows. However it was for the previous git version but i think that most of the steps (related only to the lib and not to nim/d bindings) still apply.

You can see it [here](https://www.youtube.com/watch?v=SORBYN7vOxQ)

**Comment by bubnenkoff** [comment]
Big thanks for detail answer. I will try to do it.

**Comment by bubnenkoff** [comment]
Can I use qt creator for creation GUI and than simple link with DOtherSide?
I have never wrote GUI apps...

**Comment by filcuc** [comment]
Can you explain to me better what you're trying to do? Which languages you want to use for developing your app and how you want to use libDOtherSide?

**Comment by bubnenkoff** [comment]
I using D, and if you would able provide me ready to use build I would be very glad. If licence do not allow to allow to distribute it I will try to build all myself, but sometime for not very good programmers it's much easier to start from ready to use lib, than try to produce all steps

**Comment by bubnenkoff** [comment]
There is project that allow to build Qt with D, but it's still do not have bins for test (hope soon they will be released) https://github.com/Syniurge/Calypso 

But a lot of people now are looking for modern GUI lib for D, that will allow them to easier write code. I tried to other lib like TK, but without graphical designer it's very hard to use them. So I hope that with your lib I will able to use Qt Builder

**Comment by bubnenkoff** [comment]
I troed to build it, but I got next errors: http://img.ctrlv.in/img/15/09/24/5603ab7da23d6.png

**Comment by bubnenkoff** [comment]
So if you can please to upload in any place libs for D and Qt 5.5 32Bits.
If I right understand I would need only DOtherSide and Qt for programming in D. Or not?  Would I need to have VS?

**Comment by filcuc** [comment]
This is caused by #28

**Comment by filcuc** [comment]
Fixed in 14bf9fd

**Comment by bubnenkoff** [comment]
Yes, it's helped! Thanks!

**Comment by bubnenkoff** [comment]
But is it's possible to use Qt Creator for creation graphical apps and writing logic in D?

**Comment by filcuc** [comment]
yes sure! Write your UI (.qml files) inside the creator and logic in d.
See the contactapp example

**Comment by filcuc** [comment]
https://github.com/filcuc/dqml/tree/master/examples/contactapp/source

**Comment by bubnenkoff** [comment]
In video you had renamed output DotherSide lib ti libDotherSide.
Do I need to rename it now?
I placed this lib to sample dub project and get next error http://img.ctrlv.in/img/15/09/24/5603e5484aec6.png

It's just dub default hello world app. Now App that include qml example. 

Please could you write in docs where in Windows I should put your lib and update information about should I rename it or no.

**Comment by bubnenkoff** [comment]
Also I have got question is the DOtherSide allow me use 100% of QT functionality, or it's have some limitation?

**Comment by filcuc** [comment]
please see here,
https://github.com/filcuc/dqml/blob/master/dub.json
you should add a "lib" directive in you dub file
for linking the libDOtherSide. The name of the dll is platform dependent. On unix it's with prefixed with "lib" on windows not. Try and see if you can sort it out.

**Comment by filcuc** [comment]
Yes there're some limitations.
Right now you cannot create custom QQuickItems from D or Nim. So code a QQuickItem from D.
And i didn't add support for every datatype. However i'm willing to improve if you need support

**Comment by filcuc** [comment]
Please before starting your project please download the dqml project
and compile an example to see how it works

**Comment by bubnenkoff** [comment]
I am getting strange error http://img.ctrlv.in/img/15/09/24/5603e86bd7b2a.png 
it's look like it's can't find dub.json (I use latest version of dub).

>you should add a "lib" directive in you dub file
ok I will try do it.

>Right now you cannot create custom QQuickItems 
Ok at start I will enough predefined items) It's better to any GUI than no GUI :)

**Comment by filcuc** [comment]
as i wrote you first clone the DQml repo somewhere through git.
and compile one of the bundled examples.


2015-09-24 14:15 GMT+02:00 Dmitry Bubnenkov <notifications@github.com>:

> I am getting strange error
> http://img.ctrlv.in/img/15/09/24/5603e86bd7b2a.png
> it's look like it's can't find dub.json (I use latest version of dub).
>
> you should add a "lib" directive in you dub file
> ok I will try do it.
>
> Right now you cannot create custom QQuickItems
> Ok at start I will enough predefined items) It's better to any GUI than no
> GUI :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/filcuc/DOtherSide/issues/26#issuecomment-142912343>.
>



-- 
Filippo Cucchetto

**Comment by filcuc** [comment]
After you successfully built the bundled examples. We can move forward. You will also have a working environment :)

**Comment by bubnenkoff** [comment]
I done dub dqml fetch
Clone dqml to folder
move to "hello world" app folder
copy all libs to it
run dub and now I am getting this error message: http://img.ctrlv.in/img/15/09/24/5603ebd6cdfe5.png

**Comment by filcuc** [comment]
No.
cd "your desktop"
git clone https://github.com/filcuc/dqml.git
cd dqml
dub build

and post the error you get from this steps.
It should fail because it doesn't find the libDOtherSide.dll.
You should add the directory where you put it in you PATH environement variable.
Furthermore you should not link the .lib (cause it's not a static lib) but the .dll !!!

**Comment by bubnenkoff** [comment]
`D:\foo>git clone https://github.com/filcuc/dqml.git
Cloning into 'dqml'...
remote: Counting objects: 155, done.
remote: Total 155 (delta 0), reused 0 (delta 0), pack-reused 155Receiving object
Receiving objects:  62% (97/155)
Receiving objects: 100% (155/155), 31.67 KiB | 0 bytes/s, done.
Resolving deltas: 100% (47/47), done.
Checking connectivity... done.

D:\foo>cd dqml

D:\foo\dqml>dub build
Main package must have a binary target type, not sourceLibrary. Cannot build.`

**Comment by filcuc** [comment]
Yep i because dqml is a souce package, you should try to compile an example!
cd "your desktop"
git clone https://github.com/filcuc/dqml.git
cd dqml
cd examples
cd helloworld
dub build

if it files add libDOtherSide.dll to your PATH enviroment variable or put it on C:\Windows\system32 :D

**Comment by bubnenkoff** [comment]
DOtherSide.dll
DOtherSide.lib

to system32 then run:
D:\foo\dqml\examples\helloworld>dub build

and got:

Linking...
OPTLINK (R) for Win32  Release 8.00.17
Copyright (C) Digital Mars 1989-2013  All rights reserved.
http://www.digitalmars.com/ctg/optlink.html
DOtherSide.lib
 Warning 2: File Not Found DOtherSide.lib
.dub\build\application-debug-windows-x86-dmd_2068-4DED4435482D06B8A8EB0F88F707C5
33\helloworld.obj(helloworld)
 Error 42: Symbol Undefined _dos_qabstractlistmodel_create
.dub\build\application-debug-windows-x86-dmd_2068-4DED4435482D06B8A8EB0F88F707C5
33\helloworld.obj(helloworld)
 Error 42: Symbol Undefined _dos_qabstractlistmodel_delete
.dub\build\application-debug-windows-x86-dmd_2068-4DED4435482D06B8A8EB0F88F707C5
33\helloworld.obj(helloworld)

**Comment by bubnenkoff** [comment]
Also I tried to placed this dll near hello world app:

D:\foo\dqml\examples\helloworld>dir
 Том в устройстве D имеет метку Data
 Серийный номер тома: DA61-9F17

 Содержимое папки D:\foo\dqml\examples\helloworld

24.09.2015  17:37    <DIR>          .
24.09.2015  17:37    <DIR>          ..
24.09.2015  17:34    <DIR>          .dub
24.09.2015  15:46                43 .gitignore
24.09.2015  14:43           340 992 DOtherSide.dll
24.09.2015  14:43            11 842 DOtherSide.exp
24.09.2015  14:43         1 563 868 DOtherSide.ilk
24.09.2015  14:43            20 522 DOtherSide.lib
24.09.2015  14:43         3 330 048 DOtherSide.pdb
24.09.2015  15:46               326 dub.json
24.09.2015  17:34                39 dub.selections.json
24.09.2015  15:46    <DIR>          source
               8 файлов      5 267 680 байт
               4 папок  118 642 995 200 байт свободно

D:\foo\dqml\examples\helloworld>dub
Performing "debug" build using dmd for x86.
helloworld ~master: building configuration "application"...
Linking...
OPTLINK (R) for Win32  Release 8.00.17
Copyright (C) Digital Mars 1989-2013  All rights reserved.
http://www.digitalmars.com/ctg/optlink.html
DOtherSide.lib
 Error 43: Not a Valid Library File
--- errorlevel 1
dmd failed with exit code 1.

**Comment by filcuc** [comment]
i tried this evening..the reason it's that DMD on windows use OPTLINK has a linker. This is turn is a shit. Now i'm trying to see if with GDC (so with Mingw32) the things gets solved

**Comment by bubnenkoff** [comment]
Maybe its better to ask on D forum?

**Comment by filcuc** [comment]
go on if you want. i'll try to give a go with mingw and ping you back

**Comment by bubnenkoff** [comment]
Ok, but if you will be able to do it please give me bins. My email is
bubnenkoff@gmail.com

**Comment by filcuc** [comment]
ok i was able to build the examples by using mingw32 and converting the .dll format for making it compatible with the shitty linker used on  windows by dmd.
Tomorrow evening i'll write the complete procedure.

**Comment by bubnenkoff** [comment]
ok, I am waiting for build instruction

**Comment by filcuc** [comment]
i wrote a step by step tutorial that you can read on the dqml repository, [here](https://github.com/filcuc/dqml/blob/master/WindowsUsage.md)
Let me know if it works for you

**Comment by bubnenkoff** [comment]
Cool! It's work!!! I build examples!
Could you make short video or several screenshots about building Apps in QT
creator?

**Comment by filcuc** [comment]
Sorry
but i think that's out of scope. I don't have time for that :)
You can find a lot of resources on qml.
You can start by reading the qml sections here [here](
http://qmlbook.github.io/)


2015-09-26 20:14 GMT+02:00 Dmitry Bubnenkov <notifications@github.com>:

> Cool! It's work!!! I build examples!
> Could you make short video or several screenshots about building Apps in QT
> creator?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/filcuc/DOtherSide/issues/26#issuecomment-143477201>.
>



-- 
Filippo Cucchetto

**Comment by filcuc** [comment]
Given that i'm available if you encounter bugs or need to improve the
supported datatypes.
Finally i'm happy that it worked for you too.
As you see windows it's not a developer friendly environment :)
I encourage you to always create .bat files for your build environements
(also with other projects) and to never put anything in you $PATH global
variable.


2015-09-26 23:56 GMT+02:00 Filippo Cucchetto <filippocucchetto@gmail.com>:

> Sorry
> but i think that's out of scope. I don't have time for that :)
> You can find a lot of resources on qml.
> You can start by reading the qml sections here [here](
> http://qmlbook.github.io/)
>
>
> 2015-09-26 20:14 GMT+02:00 Dmitry Bubnenkov <notifications@github.com>:
>
>> Cool! It's work!!! I build examples!
>> Could you make short video or several screenshots about building Apps in
>> QT
>> creator?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/filcuc/DOtherSide/issues/26#issuecomment-143477201>.
>>
>
>
>
> --
> Filippo Cucchetto
>



-- 
Filippo Cucchetto
